app/controllers/incidents_controller.rb

```diff
@@ -36,6 +38,18 @@ def index
 
   def show
     set_incident
+
+    # Check visibility and redirect based on user role
+    if @incident.visibility == 'draft' && !current_user&.can_manage_incidents?
+      redirect_to_root_unless_user(:can_manage_incidents?)
+      return
+    end
+
+    if @incident.visibility == 'staff' && !current_user&.staff?
+      redirect_to_root_unless_user(:staff?)
+      nil
+    end
+
     unless @incident.resolved?
       redirect_to_root_unless_user(:can_manage_incidents?)
     end
```

Comment:
This existing check should probably be "collapsed" into the two new if-checks -- that is, the visibility flag should be the indicator for access now.

(I'm assuming that once an incident is resolved, its visibility is set to public accordingly)